[#67453] added hierarchical lists to project custom fields

Kharonus opened the pull request:
# Ticket
[#67453](https://community.openproject.org/work_packages/67453)

# What are you trying to accomplish?
- added scored list and hierarchy to project attributes
- added new controller and routes
- amended view templates

Kharonus committed:
Merge branch 'dev' into implementation/67453-enable-the-custom-field-formats-for-project-attribute-usage

Kharonus committed:
[#67453] added hierarchical lists to project custom fields

- https://community.openproject.org/work_packages/67453
- added scored list and hierarchy to project attributes
- added new controller and routes
- amended view templates

NobodysNightmare commented on app/views/admin/settings/project_custom_fields/hierarchy/items/index.html.erb:
```diff
@@ -0,0 +1,34 @@
+
+<%= render(Settings::ProjectCustomFields::EditFormHeaderComponent.new(custom_field: @custom_field, selected: :items)) %>
+
+<%= render(Admin::CustomFields::Hierarchy::ItemsComponent.new(item: @active_item)) %>
```
:red_circle: The list of items still partially overlaps the tabbed navigation:

<img width="1011" height="593" alt="image" src="https://github.com/user-attachments/assets/4688d765-ac1f-4afd-9b29-0c43c3b7e75b" />

I'd have expected that we fix this before merging.

(Not quite sure whether this is exactly the right code line to put the comment)

Kharonus replied:
We can't fix this easily, as it is a misbehaviour in the view component itself. I wrote a [bug ticket](https://community.openproject.org/wp/67625) for that.

NobodysNightmare replied:
So this will block the removal of the feature flag, but it will not block this PR anymore :tada:

NobodysNightmare commented:
This PR seems to include a commit that I'd have attributed to a different branch/feature:

<img width="474" height="123" alt="image" src="https://github.com/user-attachments/assets/bfcf0bf3-1ebd-48a8-a745-fe0edc376b36" />

Can you make sure that everything is rebased properly?

NobodysNightmare reviewed (changes requested):
Note: I am expecting that this PR only covers the first AC of the corresponding feature request:

> When creating a new project attribute, an admin can choose the types "Scored List"  and "Hierarchy".

What's the plan for feature flagging this? Should the CF type availability be feature flagged away for the start or only the usage of the tree view to select a CF value? (I am assuming a feature flag, because the single-select requirement will not be fulfillable with the current tree view)

LGTM with the following exceptions (see my other comments):
* I am not sure whether the included commits are correct
* visual glitch when displaying items of the hierarchy

Kharonus commented:
> This PR seems to include a commit that I'd have attributed to a different branch/feature:
> 
> <img alt="image" width="474" height="123" src="https://private-user-images.githubusercontent.com/453584/492166257-bfcf0bf3-1ebd-48a8-a745-fe0edc376b36.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg1NDc5NzcsIm5iZiI6MTc1ODU0NzY3NywicGF0aCI6Ii80NTM1ODQvNDkyMTY2MjU3LWJmY2YwYmYzLTFlYmQtNDhhOC1hNzQ1LWZlMGVkYzM3NmIzNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkyMlQxMzI3NTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02OWQ3NDIzYTkwNTQ3YWNhOWNmOWM2OWUzYjhlMDk1NWY5OGU2M2Q1NTA4MTFmYWU0YjI0MGY2ZmY2ZDc4NmU4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.8hPDaPz3gyym4CafKLiHn_HeVcTCx6SozNEsW7XdstM">
> Can you make sure that everything is rebased properly?

Yes, I drafted this PR mainly so that @EinLama can use it. It contains the previous work, too. And once it is ready for review, I'll make sure only the relevant commits are in.

NobodysNightmare committed:
Sort hierarchy items by the name they are displayed with

The lists were a kind of make-shift solution to allow
using hierarchy custom fields in all of our UIs. However,
the sort order coming out of the service was rather useless
to a user. By sorting by the ancestry path, we sort the list by
the string that's later visible to the user, thus making
it easier to navigate the corresponding dropdown.

Kharonus committed:
Merge branch 'dev' into implementation/67453-enable-the-custom-field-formats-for-project-attribute-usage

NobodysNightmare commented:
> And once it is ready for review

I thought that I received a request for review this morning, which is why I started commenting xD

edit:

<img width="619" height="54" alt="image" src="https://github.com/user-attachments/assets/ad6beb32-0904-477d-8505-7662b0782e8f" />

NobodysNightmare committed:
Make result order returned by get_descendants deterministic

Previously the results were not returned in any specific order,
making the result not immediately usable in flat lists of the hierarchy.

By preordering the result, it's not necessary anymore to apply later
alphabetical ordering or similar. This is preferable, because alphabetical
ordering was only a workaround that didn't take specific ordering established
by the user into account.

NobodysNightmare committed:
Merge pull request #20389 from opf/select-scored-project-cf

Sort hierarchy items deterministically

Kharonus committed:
[#67453] disable hierarchy for project custom fields

- as long as the feature flag for scored lists are in place

NobodysNightmare reviewed (approved):
Looks good to me.
